tesseract_process_managers/src/task_generators/time_optimal_trajectory_generation_task_generator.cpp

```diff
@@ -0,0 +1,153 @@
+  : TaskGenerator(std::move(name))
+{
+  // Register default profile
+  composite_profiles["DEFAULT"] = std::make_shared<TimeOptimalTrajectoryGenerationProfile>();
+}
+
+int TimeOptimalTrajectoryGenerationTaskGenerator::conditionalProcess(TaskInput input, std::size_t unique_id) const
+{
+  if (input.isAborted())
+    return 0;
+
+  auto info = std::make_shared<TimeOptimalTrajectoryGenerationTaskInfo>(unique_id, name_);
+  info->return_value = 0;
+  input.addTaskInfo(info);
+
+  // --------------------
+  // Check that inputs are valid
+  // --------------------
+  Instruction* input_results = input.getResults();
+  if (!isCompositeInstruction(*input_results))
+  {
+    CONSOLE_BRIDGE_logError("Input results to TOTG must be a composite instruction");
+    return 0;
+  }
+
+  auto* ci = input_results->cast<CompositeInstruction>();
+  const ManipulatorInfo& manip_info = ci->getManipulatorInfo();
+  const auto fwd_kin = input.env->getManipulatorManager()->getFwdKinematicSolver(manip_info.manipulator);
+
+  // Get Plan Profile
+  std::string profile = ci->getProfile();
+  profile = getProfileString(profile, name_, input.composite_profile_remapping);
+  auto cur_composite_profile = getProfile<TimeOptimalTrajectoryGenerationProfile>(
+      profile, composite_profiles, std::make_shared<TimeOptimalTrajectoryGenerationProfile>());
+  if (!cur_composite_profile)
+    cur_composite_profile = std::make_shared<TimeOptimalTrajectoryGenerationProfile>();
+
+  // Create data structures for checking for plan profile overrides
+  auto flattened = flatten(*ci, moveFilter);
+  if (flattened.empty())
+  {
+    CONSOLE_BRIDGE_logWarn("TOTG found no MoveInstructions to process");
+    info->return_value = 1;
+    return 1;
+  }
+
+  double velocity_scaling_factors = cur_composite_profile->max_velocity_scaling_factor;
+  double acceleration_scaling_factors = cur_composite_profile->max_acceleration_scaling_factor;
+
+//  // Loop over all PlanInstructions
+//  for (Eigen::Index idx = 0; idx < static_cast<Eigen::Index>(flattened.size()); idx++)
+//  {
+//    profile = flattened[static_cast<std::size_t>(idx)].get().cast_const<MoveInstruction>()->getProfile();
+
+//    // Check for remapping of the plan profile
+//    std::string remap = getProfileString(profile, name_, input.plan_profile_remapping);
+//    auto cur_composite_profile = getProfile<TimeOptimalTrajectoryGenerationProfile>(
+//        remap, composite_profiles, std::make_shared<TimeOptimalTrajectoryGenerationProfile>());
+
+//    // If there is a move profile associated with it, override the parameters
+//    auto it = move_profiles.find(profile);
+//    if (it != move_profiles.end())
+//    {
+//      velocity_scaling_factors[idx] = it->second->max_velocity_scaling_factor;
+//      acceleration_scaling_factors[idx] = it->second->max_acceleration_scaling_factor;
+//    }
+//  }
+  /// @todo This algorithm does not currently support different scalings for different parts of the trajectory
```

Comment:
Do you think this is possible implement later?

Reply:
Maybe. However, I think we may be better off implementing that ourselves as a post processing step just scaling everything by the factor.

Reply:
This is currently prevented by the fact that this algorithm results in a flattened composite. I will look into preserving the subcomposite structure. That is marked as another todo.

Reply:
This is currently prevented by the fact that this algorithm results in a flattened composite. I will look into preserving the subcomposite structure. That is marked as another todo.